scripts/run_mock_server.py

```diff
@@ -0,0 +1,398 @@
+
+
+def response_decorator(primary_ws_response: bool = False) -> Callable:
+    """Factory to decorate a response handler."""
+
+    def decorator(orig_func: Callable) -> Callable:
+        """Decorate a response handler."""
+
+        @wraps(orig_func)
+        async def wrapper(
+            self: "MockZwaveJsServer", request: web_request.Request
+        ) -> web.WebSocketResponse:
+            """Wrap."""
+            logging.error(self)
+            ws_resp = web.WebSocketResponse(autoclose=False)
+            if primary_ws_response:
+                setattr(self, "primary_ws_resp", ws_resp)
+            await ws_resp.prepare(request)
+
+            if primary_ws_response:
+                version_info: VersionInfoDataType = self.network_state_dump[0]
+                # adjust min/max schemas if needed to get things to work
+                if MAX_SERVER_SCHEMA_VERSION > version_info["maxSchemaVersion"]:
+                    version_info["maxSchemaVersion"] = MAX_SERVER_SCHEMA_VERSION
+                if MIN_SERVER_SCHEMA_VERSION < version_info["minSchemaVersion"]:
+                    version_info["minSchemaVersion"] = MIN_SERVER_SCHEMA_VERSION
+                await self.send_json(version_info)
+
+            async for msg in ws_resp:
+                if msg.type == WSMsgType.TEXT:
+                    logging.debug("Message received: %s", msg.data)
+                    if msg.data == "close":
+                        await ws_resp.close()
+                    elif msg.data == "error":
+                        logging.warning("Error from client: %s", msg.data)
+
+                    try:
+                        if len(msg.data) > SIZE_PARSE_JSON_EXECUTOR:
+                            data: Union[
+                                dict, list
+                            ] = await asyncio.get_event_loop().run_in_executor(
+                                None, msg.json
+                            )
+                        else:
+                            data = msg.json()
+                    except ValueError as err:
+                        raise ExitException(
+                            f"Received invalid JSON {msg.data}"
+                        ) from err
+
+                    await orig_func(self, data)
+                elif msg.type == WSMsgType.ERROR:
+                    logging.error(
+                        "Connection closed with exception %s",
+                        ws_resp.exception(),
+                    )
+
+            logging.info("Connection closed")
+
+            return ws_resp
+
+        return wrapper
+
+    return decorator
+
+
+class MockZwaveJsServer:
+    """
+    Class to represent a mock zwave-js-server instance.
+
+    Only designed to handle one client at a time.
+    """
+
+    def __init__(
+        self,
+        network_state_dump: List[dict],
+        events_to_replay: List[dict],
+        command_responses: DefaultDict[str, DefaultDict[HashableDict, list]],
```

Comment:
events come from zwave-js whereas commands are initiated by the client, aka the lib or HA. It only makes sense to emulate a response when it's initiated by the lib or HA. In the case of the lib, we could theoretically use this in tests, in which case a list would work because we are reproducing the same behavior over and over again. But if this is to help one of us troubleshoot an issue using the HA frontend, I didn't want to force us to have to perform the exact sequence in the order that the user did in order to conduct the test. This still preserves order but allows us flexibility in case some commands are no-op to the problem, or the order doesn't matter.

With all of that being said, I don't know why I went for a nested dictionary when a single dictionary would have worked. I will refactor accordingly

Reply:
Ok. Yeah, if it's not usable in the manual test it won't matter how close we are to the real case in the test. I guess we need to have a balance.